src/transform/src/semijoin_idempotence.rs

```diff
@@ -0,0 +1,421 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Remove semijoins that are applied multiple times to no further effect.
+//!
+//! If we semijoin `A` by some of its columns against some distinct values,
+//! and then again semijoin `A` by those columns against the distinct column values
+//! that resulted, we can remove the outer semijoin.
+//!
+//! We intentionally restrict to column equivalences as we are only able to
+//! determine uniqueness of column-based keys. In principle this could be of
+//! arbitrary expressions, with a non-trivial amount of work done elsewhere.
+use std::collections::BTreeMap;
+
+use crate::TransformArgs;
+
+use mz_expr::Id;
+use mz_expr::JoinInputMapper;
+use mz_expr::MirRelationExpr;
+use mz_expr::MirScalarExpr;
+
+/// Remove redundant semijoin operators
+#[derive(Debug)]
+pub struct SemijoinIdempotence;
+
+impl crate::Transform for SemijoinIdempotence {
+    #[tracing::instrument(
+        target = "optimizer"
+        level = "trace",
+        skip_all,
+        fields(path.segment = "semijoin_idempotence")
+    )]
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        // Iteratively descend the expression, at each node attempting to simplify a join.
+        let mut bindings = BTreeMap::<Id, Vec<Replacement>>::new();
+        let mut worklist = vec![relation];
+        while let Some(expr) = worklist.pop() {
+            match expr {
+                MirRelationExpr::Let { id, value, .. } => {
+                    bindings.insert(Id::Local(*id), semijoin_replacements(&*value, &bindings));
+                }
+                MirRelationExpr::Join {
+                    inputs,
+                    equivalences,
+                    implementation,
+                    ..
+                } => {
+                    attempt_join_simplification(inputs, equivalences, implementation, &bindings);
+                }
+                _ => {}
+            }
+
+            // Continue to process the children.
+            worklist.extend(expr.children_mut());
+        }
+
+        Ok(())
```

Comment:
For the tracing part, in addition to the `tracing::instrument` attribute at the top, the body of the `transform` method needs to report the final result with a 
```rust
mz_repr::explain_new::trace_plan(&*relation);
```
call just before returning the `Result`.

Reply:
Are there some docs I can read about what is needed and how it works? I can copy/paste stuff I don't understand, but I can also see that going wrong. :D

Reply:
Yes, the API docs for [`trace_plan`](https://dev.materialize.com/api/rust-private/mz_repr/explain_new/tracing/fn.trace_plan.html) have more details. TL;DR: a high-level structure of the form:

```rust
impl crate::Transform for MyTransform {
    #[tracing::instrument(
        target = "optimizer"
        level = "trace",
        skip_all,
        fields(path.segment = "my_transform")
    )]
    fn transform(
        &self,
        expr: &mut MirRelationExpr,
        _: TransformArgs,
    ) -> Result<(), TransformError> {
        let result = ... // get the result somewhow
        mz_repr::explain_new::trace_plan(&*expr);
        result
    }
}
```

with the enclosing `tracing::instrument` attribute and the `mz_repr::explain_new::trace_plan` call just before the end is all you need.

Reply:
This resulted in a somewhat silly change: needing to add a reborrow of `relation` and scope so that the reborrow could drop. Is this not a call that should live outside the `transform` call instead at the call site? That's where I usually put my manual debugging printlns.

Reply:
The `trace_plan` call needs to be part of the enclosing span because we need to capture the end-state of the transformed relation while just before exiting that span.

If the `Transform` API was taking ownership of the input `MirRelationExpr` and returning the modified `MirRelationExpr` as part of the result (rather than borrowing), it would have been able to hide this `trace_plan` as part of a custom macro that is used instead of `tracing::instrument`, but this is not the case and I didn't to introduce too many changes in the code just for that reason.

I think the issue here is that you put the borrow in
```rust
let mut worklist = vec![&mut *relation];
```
which should be dropped by the time 
```rust
mz_repr::explain_new::trace_plan(&*expr);
```
is called. Most of the other `Transform` implementations have a complementary `action` method, which either:

1. drives the recursion by itself and is called directly from `transform` (as in `CanonicalizeMfp`), or
2. is passed as an argument to a `Visit` method on the input relation (as in `TopKElision`).

As you are not relying on a `Visit` method, the only think that I can suggest is to align with that pattern and move the code within the designated `{ ... }` lifespan to a dedicated `action` method.


```rust
impl crate::Transform for SemijoinIdempotence {
    #[tracing::instrument(
        target = "optimizer"
        level = "trace",
        skip_all,
        fields(path.segment = "semijoin_idempotence")
    )]
    fn transform(
        &self,
        relation: &mut MirRelationExpr,
        _: TransformArgs,
    ) -> Result<(), crate::TransformError> {
        action(relation)?;
        mz_repr::explain_new::trace_plan(&*relation);
        Ok(())
    }
}

fn action(relation: &mut MirRelationExpr) -> Result<(), crate::TransformError> {
    // Iteratively descend the expression, at each node attempting to simplify a join.
    let mut bindings = BTreeMap::<Id, Vec<Replacement>>::new();
    let mut worklist = vec![&mut *relation];
    while let Some(expr) = worklist.pop() {
        match expr {
            MirRelationExpr::Let { id, value, .. } => {
                bindings.insert(Id::Local(*id), semijoin_replacements(&*value, &bindings));
            }
            MirRelationExpr::Join {
                inputs,
                equivalences,
                implementation,
                ..
            } => {
                attempt_join_simplification(inputs, equivalences, implementation, &bindings);
            }
            _ => {}
        }

        // Continue to process the children.
        worklist.extend(expr.children_mut());
    }

    Ok(())
}
```

Reply:
BTW, I am able to run 
```bash
bin/environmentd --reset
``` 
on the original version with the extra `trace_plan` call (see below) without borrow checker issues. What is the error message that you are getting?

---
```rust
impl crate::Transform for SemijoinIdempotence {
    #[tracing::instrument(
        target = "optimizer"
        level = "trace",
        skip_all,
        fields(path.segment = "semijoin_idempotence")
    )]
    fn transform(
        &self,
        relation: &mut MirRelationExpr,
        _: TransformArgs,
    ) -> Result<(), crate::TransformError> {
        // Iteratively descend the expression, at each node attempting to simplify a join.
        let mut bindings = BTreeMap::<Id, Vec<Replacement>>::new();
        let mut worklist = vec![&mut *relation];
        while let Some(expr) = worklist.pop() {
            match expr {
                MirRelationExpr::Let { id, value, .. } => {
                    bindings.insert(Id::Local(*id), semijoin_replacements(&*value, &bindings));
                }
                MirRelationExpr::Join {
                    inputs,
                    equivalences,
                    implementation,
                    ..
                } => {
                    attempt_join_simplification(inputs, equivalences, implementation, &bindings);
                }
                _ => {}
            }

            // Continue to process the children.
            worklist.extend(expr.children_mut());
        }
        mz_repr::explain_new::trace_plan(&*relation);
        Ok(())
    }
}
```